PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h

```diff
@@ -349,6 +350,72 @@ class SimpleFlatTableProducer : public SimpleFlatTableProducerBase<T, edm::View<
   std::vector<std::unique_ptr<ExtVariable<T>>> extvars_;
 };
 
+template <typename T>
+class BXVectorSimpleFlatTableProducer : public SimpleFlatTableProducerBase<T, BXVector<T>> {
+public:
+  BXVectorSimpleFlatTableProducer(edm::ParameterSet const &params)
+      : SimpleFlatTableProducerBase<T, BXVector<T>>(params),
+        maxLen_(params.existsAs<unsigned int>("maxLen") ? params.getParameter<unsigned int>("maxLen")
+                                                        : std::numeric_limits<unsigned int>::max()),
+        cut_(params.getParameter<std::string>("cut"), true),
+        minBX_(params.getParameter<int>("minBX")),
+        maxBX_(params.getParameter<int>("maxBX")),
+        bxVarName_("bx") {
+    edm::ParameterSet const &varsPSet = params.getParameter<edm::ParameterSet>("variables");
+    auto varNames = varsPSet.getParameterNamesForType<edm::ParameterSet>();
+    if (std::find(varNames.begin(), varNames.end(), bxVarName_) != varNames.end()) {
+      throw cms::Exception("Configuration",
+                           "BXVectorSimpleFlatTableProducer already defines the " + bxVarName_ +
+                               "internally and thus you should not specify it yourself");
+    }
+  }
+
+  static void fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
+    edm::ParameterSetDescription desc = SimpleFlatTableProducerBase<T, BXVector<T>>::baseDescriptions();
+    desc.add<std::string>("cut", "")->setComment(
+        "selection on the main input collection (but selection can not be bx based)");
+    desc.addOptional<unsigned int>("maxLen")->setComment(
+        "define the maximum length of the input collection to put in the branch");
+    desc.add<int>("minBX", -2)->setComment("min bx (inclusive) to include");
+    desc.add<int>("maxBX", 2)->setComment("max bx (inclusive) to include");
+    descriptions.addWithDefaultLabel(desc);
+  }
+
+  std::unique_ptr<nanoaod::FlatTable> fillTable(const edm::Event &iEvent,
+                                                const edm::Handle<BXVector<T>> &prod) const override {
+    std::vector<const T *> selObjs;
+    std::vector<int> selObjBXs;
+
+    if (prod.isValid() || !(this->skipNonExistingSrc_)) {
+      const int minBX = std::max(minBX_, prod->getFirstBX());
+      const int maxBX = std::min(maxBX_, prod->getLastBX());
+      for (int bx = minBX; bx <= maxBX; bx++) {
+        for (size_t objNr = 0, nrObjs = prod->size(bx); objNr < nrObjs; ++objNr) {
+          const auto &obj = prod->at(bx, objNr);
+          if (cut_(obj)) {
+            selObjs.push_back(&obj);
+            selObjBXs.push_back(bx);
+          }
+          if (selObjs.size() >= maxLen_)
+            break;
+        }
+      }
+    }
+    auto out = std::make_unique<nanoaod::FlatTable>(selObjs.size(), this->name_, false, this->extension_);
+    for (const auto &var : this->vars_)
+      var->fill(selObjs, *out);
+    out->template addColumn<int>(bxVarName_, selObjBXs, "BX of the L1 candidate");
```

Comment:
Even if that won't happen in practice, it would be nice if that column was only added if `minBX != maxBX`

Reply:
so on this, I would like to offer reasoning for always having it in the interests of usability. 

For me, it is much easier to know that bx always exist so I can just do bx==0 rather than trying to figure out if the bx variable exists and if so then do bx==0

I think we make things slightly harder for ourselves to have a variable which sometimes exists and sometimes doesnt given the cost of storing said variable is effectively zero as whenever we wouldnt store it would be zero. 

However given this is minor, if the above argument doesnt convince you, how about I make it a configurable parameter whether to store the bx when there is only one so we have this option if we dont store all bx in the future. How does that sound?

Reply:
From experience: it is always better to explicitly require e.g. `BX=0` to avoid cases where the file suddenly has more or less BX stored. (it came several times by surprise that the re-emulated objects are only BX0 and thus had seemingly lower occupancy 😅 )
It could be just the smallest possible type, i.e. int8 or so, right?

Reply:
That's all right, I won't insist on it, given that in practice we will store all BXs for central prompt data production. Indeed it can be an int8 (just be aware that int8 is interpreter as a character and sometimes behaves weirdly with ROOT plotting, to avoid that an int16 is also fine).